File: C++/Graphs/Problem6.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
// Distance of Nearing cell having 1
class Solution{
public:
  vector<vector<int>> nearestCell(vector<vector<int>> grid){
    int n = grid.size();
    int m = grid[0].size();
    vector<vector<int>> vis(n, vector<int>(m,0));
    vector<vector<int>> ans(n, vector<int>(m, 0));
    queue<pair<pair<int,int>,int>> q;
    for(int i=0; i<n; i++){
      for(int j=0; j<m; j++){
        if(grid[i][j]==1){
          q.push({{i,j}, 0});
          vis[i][j] = 1;
        }
      }
    }
    int delrow[] = {-1, 0, 1, 0};
    int delcol[] = {0, -1, 0, 1};
    while(!q.empty()){
      int row = q.front().first.first;
      int col = q.front().first.second;
      int steps = q.front().second;
      q.pop();

      ans[row][col] = steps;

      for(int i=0; i<4; i++){
        int newr = delrow[i] + row;
        int newc = delcol[i] + col;
        if(newr>=0 && newr<n && newc>=0 && newc<m && !vis[newr][newc]){
          vis[newr][newc] = 1;
          q.push({{newr,newc}, steps+1});
        }
      }
    }
    return ans;
  }
};

int main(){
  vector<vector<int>> grid = {{0,1,1,0},{1,1,0,0},{0,0,1,1}};
  Solution obj;
  vector<vector<int>> ans;
  cout << "old grid: " << endl;
  for(int i=0; i<grid.size(); i++){
    for(int j=0; j<grid[i].size(); j++){
      cout << grid[i][j] << " ";
    }
    cout << endl;
  }

  cout << "New grid with distance of nearest cell having 1: " << endl;
  ans = obj.nearestCell(grid);
  for(int i=0; i<ans.size(); i++){
    for(int j=0; j<ans[i].size(); j++){
      cout << ans[i][j] << " ";
    }
    cout << endl;
  }

  return 0;
}

File: JSCheatSheet/Graphs/Problem6.js
class Solution{
  nearestCell(grid){
    let n = grid.length;
    let m = grid[0].length;
    const vis = Array.from({length: n}, () => Array(m).fill(0));
    const ans = Array.from({length: n}, () => Array(m).fill(0));
    const q = []
    for (let row = 0; row < grid.length; row++) {
      for (let col = 0; col < grid[row].length; col++) {
        if (grid[row][col] === 1) {
            q.push([row, col, 0]);
            vis[row][col] = 1;
        }
      }
    }
    const delrow = [-1, 0, 1, 0];
    const delcol = [0, -1, 0, 1];
    while(q.length){
      const front = q.shift();
      let row = front[0];
      let col = front[1];
      let steps = front[2];

      ans[row][col] = steps;
      for(let i=0; i<4; i++){
        let newr = delrow[i] + row;
        let newc = delcol[i] + col;
        if(newr>=0 && newr<n && newc>=0 && newc<m && !vis[newr][newc]){
          vis[newr][newc] = 1;
          q.push([newr, newc, steps+1]);
        }
      }
    }
    return ans;
  }

  printGraph(grid){
    for(const row of grid){
      let r = "";
      for(const val of row){
        r += val + " "
      }
      console.log(r);
    }
  }
}

const grid = [
  [1, 0, 1],
  [1, 1, 0],
  [1 ,0, 0]
]
console.log("old grid: ");
const obj = new Solution();
obj.printGraph(grid);
console.log("new grid; ");
const ans = obj.nearestCell(grid);
obj.printGraph(ans);
